### ISSUE TITLE
getBreakendAssemblyContigBreakpointInterval() bounds do not always match

### ISSUE THREAD

**Issue by sheenams** [PapenfussLab/gridss#177]
I'm attempting to parse the output of gridss, but am running into various edge cases that I was not expecting. Can you tell me if this logic is correct:

For each gridss Event ID, there will be (at least) two calls:

```
1       11046186        gridss1_982o    G       ]1:209303978]G  99.75   LOW_QUAL;SINGLE_ASSEMBLY        AS=0;ASQ=0.00;ASRP=0;ASSR=3;BA=0;BANRP=0;BANRPQ=0.00;BANSR=0;BANSRQ=0.00;BAQ=0.00;BASRP=0;BASSR=0;BEID=asm20-1136;BEIDH=0;BEIDL=60;BQ=0.00;BSC=0;BSCQ=0.00;BUM=0;BUMQ=0.00;BVF=0;CAS=0;CASQ=0.00;CIPOS=-1,0;CIRPOS=-1,0;CQ=73.22;EVENT=gridss1_982;HOMLEN=1;HOMSEQ=C;IC=0;IHOMPOS=-1,2;IQ=0.00;PARID=gridss1_982h;RAS=1;RASQ=73.22;REF=1;REFPAIR=0;RP=0;RPQ=0.00;SB=0.5;SC=2X58M;SR=1;SRQ=26.53;SVTYPE=BND;VF=3     GT:ASQ:ASRP:ASSR:BANRP:BANRPQ:BANSR:BANSRQ:BAQ:BASRP:BASSR:BQ:BSC:BSCQ:BUM:BUMQ:BVF:CASQ:IC:IQ:QUAL:RASQ:REF:REFPAIR:RP:RPQ:SR:SRQ:VF   .:0.00:0:3:0:0.00:0:0.00:0.00:0:0:0.00:0:0.00:0:0.00:0:0.00:0:0.00:99.75:73.22:1:0:0:0.00:1:26.53:3
1       209303978       gridss1_982h    C       C[1:11046186[   99.75   LOW_QUAL;SINGLE_ASSEMBLY        AS=1;ASQ=73.22;ASRP=0;ASSR=3;BA=0;BANRP=0;BANRPQ=0.00;BANSR=0;BANSRQ=0.00;BAQ=0.00;BASRP=0;BASSR=0;BEID=asm20-1136;BEIDH=60;BEIDL=0;BQ=0.00;BSC=0;BSCQ=0.00;BUM=0;BUMQ=0.00;BVF=0;CAS=0;CASQ=0.00;CIPOS=-1,0;CIRPOS=-1,0;CQ=73.22;EVENT=gridss1_982;HOMLEN=1;HOMSEQ=C;IC=0;IHOMPOS=-1,2;IQ=0.00;PARID=gridss1_982o;RAS=0;RASQ=0.00;REF=0;REFPAIR=0;RP=0;RPQ=0.00;SB=0.5;SC=58M2X;SR=1;SRQ=26.53;SVTYPE=BND;VF=3     GT:ASQ:ASRP:ASSR:BANRP:BANRPQ:BANSR:BANSRQ:BAQ:BASRP:BASSR:BQ:BSC:BSCQ:BUM:BUMQ:BVF:CASQ:IC:IQ:QUAL:RASQ:REF:REFPAIR:RP:RPQ:SR:SRQ:VF   .:73.22:0:3:0:0.00:0:0.00:0.00:0:0:0.00:0:0.00:0:0.00:0:0.00:0:0.00:99.75:0.00:0:0:0:0.00:1:26.53:3

```

For this example, 
EVENT=gridss1_982
event1=gridss1_982o
event2=gridss1_982h

Where:
event1 POS == event2 ALT position 
11046186 == C[1:11046186[
and 
event2 POS == event1 ALT position  
209303978 == ]1:209303978]G

However, after processing a few thousand lines of calls, I've come across a couple cases where:
 1. 'event2' (the 'h' event ID) isn't found. Its listed in the data for 'event1', but doesn't have its own line

```
3       37242109        gridss53_606o   C       C[5:98216958[   225.28  LOW_QUAL;SINGLE_ASSEMBLY        AS=0;ASQ=0.00;ASRP=1;ASSR=8;BA=0;BANRP=0;BANRPQ=0.00;BANSR=0;BANSRQ=0.00;BAQ=0.00;BASRP=0;BASSR=0;BEID=asm98-1031;BEIDH=78;BEIDL=0;BQ=0.00;BSC=0;BSCQ=0.00;BUM=0;BUMQ=0.00;BVF=0;CAS=0;CASQ=0.00;CIPOS=0,22;CIRPOS=0,22;CQ=204.07;EVENT=gridss53_606;HOMLEN=22;HOMSEQ=TTTTTTTTTTTTTTTTTTTTTT;IC=0;IHOMPOS=0,22;IQ=0.00;PARID=gridss53_606h;RAS=1;RASQ=204.07;REF=2;REFPAIR=0;RP=1;RPQ=21.21;SB=0.0;SC=77M1X21N1X;SR=0;SRQ=0.00;SVTYPE=BND;VF=9      GT:ASQ:ASRP:ASSR:BANRP:BANRPQ:BANSR:BANSRQ:BAQ:BASRP:BASSR:BQ:BSC:BSCQ:BUM:BUMQ:BVF:CASQ:IC:IQ:QUAL:RASQ:REF:REFPAIR:RP:RPQ:SR:SRQ:VF   .:0.00:1:8:0:0.00:0:0.00:0.00:0:0:0.00:0:0.00:0:0.00:0:0.00:0:0.00:225.28:204.07:2:0:1:21.21:0:0.00:9
```
 2. event1 POS != event2 ALT position. The ID is the same, the event positions match the other way (event2 ALT == event1 POS). Example: 
```
7       98550671        gridss133_319o  T       T[7:98550704[   215.25  LOW_QUAL        AS=1;ASQ=94.82;ASRP=0;ASSR=6;BA=0;BANRP=0;BANRPQ=0.00;BANSR=0;BANSRQ=0.00;BAQ=0.00;BASRP=0;BASSR=0;BEID=asm133-2227,asm133-548;BEIDH=77,82;BEIDL=0,0;BQ=43.81;BSC=2;BSCQ=43.81;BUM=0;BUMQ=0.00;BVF=0;CAS=0;CASQ=0.00;CIPOS=-17,0;CIRPOS=-17,0;CQ=215.25;EVENT=gridss133_319;HOMLEN=17;HOMSEQ=TGTGTGTGTGTGTGTGT;IC=0;IHOMPOS=-17,0;IQ=0.00;PARID=gridss133_319h;RAS=1;RASQ=46.24;REF=1;REFPAIR=1;RP=0;RPQ=0.00;SB=1.0;SC=64M1X16N1X;SR=3;SRQ=74.19;SVTYPE=BND;VF=5   GT:ASQ:ASRP:ASSR:BANRP:BANRPQ:BANSR:BANSRQ:BAQ:BASRP:BASSR:BQ:BSC:BSCQ:BUM:BUMQ:BVF:CASQ:IC:IQ:QUAL:RASQ:REF:REFPAIR:RP:RPQ:SR:SRQ:VF   .:94.82:0:6:0:0.00:0:0.00:0.00:0:0:43.81:2:43.81:0:0.00:0:0.00:0:0.00:215.25:46.24:1:1:0:0.00:3:74.19:5
7       98550704        gridss133_319h  T       ]7:98550669]T   285.77  LOW_QUAL        AS=1;ASQ=116.76;ASRP=0;ASSR=9;BA=0;BANRP=0;BANRPQ=0.00;BANSR=0;BANSRQ=0.00;BAQ=0.00;BASRP=0;BASSR=0;BEID=asm133-2227,asm133-548;BEIDH=0,0;BEIDL=77,82;BQ=42.57;BSC=2;BSCQ=42.57;BUM=0;BUMQ=0.00;BVF=0;CAS=0;CASQ=0.00;CIPOS=-15,0;CIRPOS=-15,0;CQ=215.25;EVENT=gridss133_319;HOMLEN=15;HOMSEQ=TGTGTGTGTGTGTGT;IC=0;IHOMPOS=-15,0;IQ=0.00;PARID=gridss133_319o;RAS=1;RASQ=94.82;REF=7;REFPAIR=6;RP=0;RPQ=0.00;SB=1.0;SC=1X14N1X32M;SR=3;SRQ=74.19;SVTYPE=BND;VF=7    GT:ASQ:ASRP:ASSR:BANRP:BANRPQ:BANSR:BANSRQ:BAQ:BASRP:BASSR:BQ:BSC:BSCQ:BUM:BUMQ:BVF:CASQ:IC:IQ:QUAL:RASQ:REF:REFPAIR:RP:RPQ:SR:SRQ:VF   .:116.76:0:9:0:0.00:0:0.00:0.00:0:0:42.57:2:42.57:0:0.00:0:0.00:0:0.00:285.77:94.82:7:6:0:0.00:3:74.19:7
```
Can you please let me know if my assumptions are not correct and what it means when these edge cases come up. 

Thanks,
Sheena

**Comment by d-cameron** [comment]
>For each gridss Event ID, there will be (at least) two calls:

With the release of GRIDSS 2, this is no longer the case as GRIDSS now calls single breakend variants and these don't have a partner.

That said, yes: all breakpoint calls should have two records and those records should match. They should have matching QUAL, matching positions, and all fields that are not specific to that breakend should have the same values.

Unfortunately, this is not always the case. Whilst there are indeed a few bugs in some edge cases remaining, the vast majority of the inconsistent calls are due to issues in the underlying data. Since GRIDSS performs operations in a streaming manner, the two sides of a breakpoint are calculated independently. If the data doesn't match/make sense, the resultant call will have issues.

For example, if a mark duplicates algorithm that only considers the primary read alignment (pretty much all of them) is run to remove duplicate reads, then you'll end up with orphan supplementary alignment records that have no primary alignment record. The SA tag will still be there so at the supplementary alignment position GRIDSS will count it as a split read, but at the primary alignment position, there is no record so GRIDSS doesn't score that split read. If the QUAL calculated at the primary alignment position is too low, no call will be made thus the orphan BND records (example 1). If the record being included/excluded causes a different microhomology length to be calculated, then the called positions may not match (example 2).

That said, the GRIDSS preprocessing step fixes most of the common data issues, and your examples look a lot like instances of #176, or #167. If you are able send through a sliced assembly.sv.bam in the assembly.gridss.working directory that contains only the assembly contigs +-2,000bp of the positions in your examples, I can take a look to see exactly what's going on and create a fix if it is indeed a problem with GRIDSS.

TDLR: different QUAL scores indicate either a) a data consistency issue that I don't yet handle in the preprocessing step, or b) a bug in GRIDSS.

**Comment by sheenams** [comment]
The first example is an issue with my data (the second read was filtered out). I've attached the file with the surrounding reads for the second example. 
[gridss_assembly.sam.gz](https://github.com/PapenfussLab/gridss/files/2502214/gridss_assembly.sam.gz)

Thanks for your help

**Comment by d-cameron** [comment]
Ok, I've finally managed to reproduce the error from your attached file (thanks). The issue is due to the different interpretation of the assembly bounds in the presence of repeats. It looks like an off-by-one error but I need to dig further to work out the root cause because the simpler test cases I created are all passing. Thanks for reporting this.

**Comment by d-cameron** [comment]
In terms of the output, the reason the BND positions disagree is because they're using different assemblies as the nominal breakpoint. This is, in turn, due to the higher quality assembly being considered of lower quality only at one breakend. This is due to the two sides of the assembly disagreeing on the assembly interval by 1bp. I still need to track down whether this is due inconsistent rounding up/down of the homology interval, or another underlying cause.

TLDR: both calls (15 or 16 bp homology at the breakpoint) have support, GRIDSS is just not choosing the same one on both sides.

**Comment by d-cameron** [comment]
#266
